src/webui/api/appcontroller.cpp

```diff
@@ -1083,6 +1085,24 @@ void AppController::defaultSavePathAction()
     setResult(BitTorrent::Session::instance()->savePath().toString());
 }
 
+void AppController::getDirectoriesAction()
+{
+    requireParams({u"path"_s});
+
+    const QString path = params().value(u"path"_s);
+    if (path.isEmpty() || path.startsWith(u':'))
+        throw APIError(APIErrorType::BadParams);
+
+    const QDir dir {path};
+    if (!dir.isAbsolute())
+        throw APIError(APIErrorType::BadParams);
+    if (!dir.exists())
+        throw APIError(APIErrorType::NotFound);
```

Comment:
Here using `APIErrorType::NotFound` is wrong. It maps to http 404 error which is incorrect for this case.
Just combine it with the above:
```suggestion
    if (!dir.isAbsolute() || !dir.exists())
        throw APIError(APIErrorType::BadParams);
```

Reply:
@Chocobo1 Don't you think there should be a different response from server when we give wrong parameter (something, starting with `:` or not path at all) and when we give a proper path, but to a nonexistent location?

Reply:
> Don't you think there should be a different response from server when we give wrong parameter (something, starting with `:` or not path at all) and when we give a proper path, but to a nonexistent location?

I believe it should be. They are obviously different cases. If you don't like it to be "Not found" then it could be "Conflict".

Reply:
>I believe it should be. They are obviously different cases. If you don't like it to be "Not found" then it could be "Conflict".

'Conflict' is http error 409 which is also inappropriate.

>@Chocobo1 Don't you think there should be a different response from server when we give wrong parameter (something, starting with : or not path at all) and when we give a proper path, but to a nonexistent location?

Yes, I would still use `BadParams` (http error 400) but provide different message in `message` field:
https://github.com/qbittorrent/qBittorrent/blob/4076712911349feb51eeacdb3ff63516bbc26b59/src/webui/api/apierror.h#L45

Reply:
> Yes, I would still use `BadParams` (http error 400) but provide different message in `message` field

But "Bad params" is inappropriate in this case, isn't it? Parameter is valid.

Reply:
>[400 Bad Request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400)
The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).

Reply:
>[404 Not Found](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404)
The server cannot find the requested resource. In the browser, this means the URL is not recognized. In an API, this can also mean that the endpoint is valid but *the resource itself does not exist*.

Isn't it the case?

Reply:
Based on these descriptions:
https://www.rfc-editor.org/rfc/rfc9110#name-client-error-4xx
https://www.rfc-editor.org/rfc/rfc9110#name-server-error-5xx
I think `if (!dir.exists())` should return 5xx error code. Maybe 500 would suffice?

Reply:
Or if we still can't agree on 4xx or 5xx. Then maybe just return `200 OK` with an empty list.

Reply:
> Then maybe just return `200 OK` with an empty list for `!dir.exists()` case.

This looks confusing, as this response is expected for an existing empty directory. 

>Maybe error 500 would suffice?

500 code with the corresponding message looks acceptable to me.

Reply:
**To sum up**. You think the response codes should look like this:

- Proper path returns `200` (even when directory is empty)
- Not path (string which is not a path) returns `400` (Bad Request)
- Path to a non-existent directory  `500` (Internal Server Error)

@Chocobo1 

In my opinion, returning 500 is unacceptable and messy, because it means, that something went wrong/we have unhandled error. This is a bit too generic crash response.    
I think `404` (Not Found) is the best option for this case, because we make a query to a non-existent directory. So in this case this is the best option, because server did not find a directory to make query.

> The 404 (Not Found) status code indicates that the origin server did not find a current representation for the [target resource](https://www.rfc-editor.org/rfc/rfc9110#target.resource) or

> The HTTP 404 Not Found response status code indicates that the server cannot find the requested resource

 Implementation on the client side would be also more readable, when we handle not found than exception

```js
if (code == HttpStatusCode.NOT_FOUND){
console.log("this directory does not exist")
//VS
if (code == HttpStatusCode.INTERNAL_SERVER_ERROR){
console.log("this directory does not exist, or sever crashed IDK")
   if (message == "directory does not exist"){
      console.log("now we know, that this dir does not exist")
```

Reply:
>I think 404 (Not Found) is the best option for this case, because we make a query to a non-existent directory. So in this case this is the best option, because server did not find a directory to make query.

I don't insist on my opinion anymore. I don't mind you use it.